File: C++/ZIMBABWE.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <cmath>
using namespace std;

using ull = unsigned long long;

ull Mapping_array(vector<int>& check) {
    ull rtr = 0;
    ull base = 1;
    for (int i = 0; i < 10; i++) {
        rtr += base * check[i];
        base *= 14;  // 16진수 기반으로 계산
    }
    return rtr;
}

ull ZIMBABWE(vector<int>& check, unordered_map<ull, vector<int>>& dp, int m, ull count) {
    if (count == 0) return 0;

    ull any = Mapping_array(check);
    if (dp.find(any) != dp.end()) {
        return any;
    }

    dp[any] = vector<int>(m, 0);
    for (int i = 0; i < 10; i++) {
        if (check[i] > 0) {
            check[i]--;
            ull cnt = i * count;
            ull tempt = cnt % m;
            count /= 10;

            ull val = ZIMBABWE(check, dp, m, count);

            if (val == 0) {
                dp[any][tempt] = 1;
            } else {
                for (int j = 0; j < m; j++) {
                    dp[any][(j + tempt) % m] = (dp[any][(j + tempt) % m] + dp[val][j]) % 1000000007;
                }
            }
            count *= 10;
            check[i]++;
        }
    }
    return any;
}

ull find(int idx, vector<int>& lst, vector<int>& check, unordered_map<ull, vector<int>>& dp, int m, ull count, unordered_map<ull, vector<int>>& dp2) {
    if (idx == lst.size()) return 0;

    ull any = Mapping_array(check);
    if (dp2.find(any) != dp2.end()) {
        return any;
    }

    dp2[any] = vector<int>(m, 0);
    for (int i = 0; i < 10; i++) {
        if (check[i] > 0 && i < lst[idx]) {
            check[i]--;
            ull cnt = i * count;
            count /= 10;
            ull val = ZIMBABWE(check, dp, m, count);

            count *= 10;
            check[i]++;
            ull tempt = cnt % m;
            if (val == 0) {
                dp2[any][tempt] = 1;
            } else {
                for (int j = 0; j < m; j++) {
                    dp2[any][(j + tempt) % m] = (dp2[any][(j + tempt) % m] + dp[val][j]) % 1000000007;
                }
            }
        } else if (check[i] > 0 && i == lst[idx]) {
            ull cnt = lst[idx] * count;
            count /= 10;
            check[i]--;

            ull val = find(idx + 1, lst, check, dp, m, count, dp2);

            count *= 10;
            ull tempt = cnt % m;
            check[i]++;
            if (val == 0) continue;

            for (int j = 0; j < m; j++) {
                dp2[any][(j + tempt) % m] = (dp2[any][(j + tempt) % m] + dp2[val][j]) % 1000000007;
            }
        }
    }
    return any;
}

int main() {
    int T;
    cin >> T;
    vector<int> answer;

    for (int t = 0; t < T; t++) {
        ull e;
        int m;
        cin >> e >> m;

        ull tmp = e;
        vector<int> lst;
        while (e > 0) {
            lst.push_back(e % 10);
            e /= 10;
        }
        reverse(lst.begin(), lst.end());

        vector<int> check(10, 0);
        unordered_map<ull, vector<int>> dp;
        unordered_map<ull, vector<int>> dp2;
        dp[0] = vector<int>(m, 0);

        for (int i : lst) {
            check[i]++;
        }

        ull rtr = find(0, lst, check, dp, m, pow(10, lst.size() - 1), dp2);
        answer.push_back(dp2[rtr][0]);
    }

    for (int k : answer) {
        cout << k << endl;
    
    }

    return 0;
}

File: PY/ZIMBABWE.py
import sys

input = sys.stdin.readline

T = int(input())
def Mapping_array(check):
    rtr = 0
    for i in range(10):
        rtr += (16**i)*check[i]
    return rtr

def ZIMBABWE(check, dp,m, count):
    if count == 0:
        return 0
    any = Mapping_array(check)
    if any in dp:
        return any
    dp[any] = [0 for k in range(m)]
    cnt = 0
    for i in range(10):
        if check[i] > 0:
            check[i] -= 1
            cnt = i*count
            tempt = cnt%m
            count //= 10
            
            val = ZIMBABWE(check, dp,m, count)
            
            if val == 0:
                dp[any][tempt] = 1
            else:
                for j in range(m):
                    dp[any][(j+tempt)%m]= (dp[any][(j+tempt)%m] + dp[val][j])%1000000007
            count *= 10
            cnt = 0
            check[i] += 1
    return any

def find(idx, lst, check, dp, m, count, dp2):
    if idx == len(lst):
        return 0
    any = Mapping_array(check)
    dp2[any] = [0 for _ in range(m)]
    for i in range(10):
        if check[i] > 0 and i < lst[idx]:
            check[i] -= 1
            cnt = i*count
            count //= 10
            val = ZIMBABWE(check, dp, m, count)
            
            count *= 10
            check[i] += 1
            tempt = cnt%m
            if val == 0:
                dp2[any][tempt] = 1
            else:
                for j in range(m):
                    dp2[any][(j+tempt)%m]= (dp2[any][(j+tempt)%m] + dp[val][j])%1000000007
        elif check[i] > 0 and i == lst[idx]:
            cnt = lst[idx]*count
            count //= 10
            check[i] -= 1
            
            val = find(idx+1, lst, check, dp, m, count, dp2)
            
            count *= 10
            tempt = cnt%m
            check[i] += 1
            if val == 0:
                continue
            else:
                for j in range(m):
                    dp2[any][(j+tempt)%m]= (dp2[any][(j+tempt)%m] + dp2[val][j])%1000000007
    return any             
answer = []
for _ in range(T):
    e, m = map(int, input().split())
    tmp = e
    lst = []
    while e > 0:
        lst.append(e%10)
        e = e//10
    lst.reverse()
    check = [0 for _ in range(10)]
    dp = dict()
    dp2 = dict()
    dp[0] = [0 for _ in range(m)]
    for i in lst:
        check[i] += 1
    rtr = find(0, lst, check, dp, m, 10**(len(lst)-1),dp2)
    answer.append(dp2[rtr][0])
    # any = ZIMBABWE(check,dp,m,10**(len(lst)-1))
    # print(dp[any][0])
for k in answer:
    print(k)